File: frontend/src/api/ai.ts
import request from '@/utils/request'

export interface AIModel {
  name: string
  size?: string
  parameterSize?: string
  family?: string
  format?: string
  modifiedAt?: string
}

export interface AIStatus {
  status: 'connected' | 'disconnected' | string
  message: string
  url?: string | null
  hint?: string | null
  localOnline: boolean
  remoteOnline: boolean
  remoteEnabled: boolean
}

export interface AIChatRequest {
  model: string
  prompt: string
  stream?: boolean
  context?: number[]
}

export const aiApi = {
  getModels: () => request.get('/ai/models'),
  getStatus: () => request.get('/ai/status'),
  chat: (data: AIChatRequest) => request.post('/ai/chat', { ...data, stream: false }),

  async chatStream(
    data: AIChatRequest,
    onMessage: (chunk: string) => void,
    onDone?: () => void,
    onError?: (err: Error) => void,
  ) {
    try {
      const token = localStorage.getItem('token')
      const res = await fetch('/api/ai/chat', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          ...(token ? { Authorization: `Bearer ${token}` } : {}),
        },
        body: JSON.stringify({ ...data, stream: true }),
      })
      if (!res.ok || !res.body) throw new Error(`AI stream failed: ${res.status}`)
      const reader = res.body.getReader()
      const decoder = new TextDecoder()
      let buffer = ''
      while (true) {
        const { done, value } = await reader.read()
        if (done) break
        buffer += decoder.decode(value, { stream: true })
        const lines = buffer.split('\n')
        buffer = lines.pop() || ''
        for (const line of lines) {
          const trimmed = line.trim()
          if (!trimmed) continue
          try {
            const json = JSON.parse(trimmed)
            if (json.response) onMessage(json.response)
            if (json.done) {
              onDone?.()
              return
            }
          } catch {
            onMessage(trimmed)
          }
        }
      }
      onDone?.()
    } catch (e) {
      onError?.(e instanceof Error ? e : new Error(String(e)))
    }
  },
}

File: api/app/schemas/ai.py
from pydantic import BaseModel, ConfigDict, Field


class AIModel(BaseModel):
    name: str
    size: int = 0
    parameter_size: str = Field("", serialization_alias="parameterSize")
    family: str = ""
    format: str = ""
    modified_at: str = Field("", serialization_alias="modifiedAt")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)


class AIModelsResponse(BaseModel):
    models: list[AIModel] = []


class AIChatRequest(BaseModel):
    model: str
    prompt: str
    stream: bool = True
    context: list[int] | None = None


class AIStatusResponse(BaseModel):
    status: str
    message: str = ""
    url: str | None = None
    hint: str | None = None
    local_online: bool = Field(False, serialization_alias="localOnline")
    remote_online: bool = Field(False, serialization_alias="remoteOnline")
    remote_enabled: bool = Field(False, serialization_alias="remoteEnabled")

    model_config = ConfigDict(populate_by_name=True, by_alias=True)
